lib/withCrop.js

```diff
@@ -0,0 +1,640 @@
+        // reason y was way off at fast speeds moving sw->ne with fixed aspect only, I couldn't
+        // figure out why.
+        if (evData.lastYCrossover === false) {
+          newY = crop.y - newSize.height;
+        } else {
+          newY = crop.y + (crop.height - newSize.height);
+        }
+      }
+
+      // Apply x/y/width/height changes depending on ordinate (fixed aspect always applies both).
+      if (crop.aspect || ReactCropHoc.xyOrds.indexOf(ord) > -1) {
+        crop.x = clamp(newX, 0, 100 - newSize.width);
+        crop.y = clamp(newY, 0, 100 - newSize.height);
+
+        crop.width = newSize.width;
+        crop.height = newSize.height;
+      } else if (ReactCropHoc.xOrds.indexOf(ord) > -1) {
+        crop.x = clamp(newX, 0, 100 - newSize.width);
+        crop.width = newSize.width;
+      } else if (ReactCropHoc.yOrds.indexOf(ord) > -1) {
+        crop.y = clamp(newY, 0, 100 - newSize.height);
+        crop.height = newSize.height;
+      }
+
+      evData.lastYCrossover = evData.yCrossOver;
+      this.crossOverCheck();
+    }
+
+    straightenYPath(clientX) {
+      const evData = this.evData;
+      const ord = evData.ord;
+      const cropOffset = evData.cropOffset;
+      const cropStartWidth = (evData.cropStartWidth / 100) * evData.imageWidth;
+      const cropStartHeight = (evData.cropStartHeight / 100) * evData.imageHeight;
+      let k;
+      let d;
+
+      if (ord === 'nw' || ord === 'se') {
+        k = cropStartHeight / cropStartWidth;
+        d = cropOffset.top - (cropOffset.left * k);
+      } else {
+        k = -cropStartHeight / cropStartWidth;
+        d = cropOffset.top + (cropStartHeight - (cropOffset.left * k));
+      }
+
+      return (k * clientX) + d;
+    }
+
+    nextCropState(crop, validateCrop) {
+      const nextCrop = { ...ReactCropHoc.defaultCrop, ...crop };
+      if (validateCrop) {
+        this.setState({ cropInvalid: isCropInvalid(nextCrop) });
+      }
+      return nextCrop;
+    }
+
+    crossOverCheck() {
+      const evData = this.evData;
+
+      if ((!evData.xCrossOver && -Math.abs(evData.cropStartWidth) - evData.xDiffPc >= 0) ||
+        (evData.xCrossOver && -Math.abs(evData.cropStartWidth) - evData.xDiffPc <= 0)) {
+        evData.xCrossOver = !evData.xCrossOver;
+      }
+
+      if ((!evData.yCrossOver && -Math.abs(evData.cropStartHeight) - evData.yDiffPc >= 0) ||
+        (evData.yCrossOver && -Math.abs(evData.cropStartHeight) - evData.yDiffPc <= 0)) {
+        evData.yCrossOver = !evData.yCrossOver;
+      }
+
+      const swapXOrd = evData.xCrossOver !== evData.startXCrossOver;
+      const swapYOrd = evData.yCrossOver !== evData.startYCrossOver;
+
+      evData.inversedXOrd = swapXOrd ? inverseOrd(evData.ord) : false;
+      evData.inversedYOrd = swapYOrd ? inverseOrd(evData.ord) : false;
+    }
+
+    render() {
+      return (<BaseComponent
+        nextCropState={this.nextCropState.bind(this)}
```

Comment:
Many of this are already bound in the constructor, and can we move the rest to the constructor too please? (for consistency but also eslint complains as it doesn't like binds in the render method) "This is bad for performance, as it will result in the garbage collector being invoked way more than is necessary. It may also cause unnecessary re-renders if a brand new function is passed as a prop to a component that uses reference equality check on the prop to determine if it should update."